src/ball_tree.jl

```diff
@@ -4,7 +4,7 @@
 # The tree uses the triangle inequality to prune the search space
 # when finding the neighbors to a point,
 struct BallTree{V <: AbstractVector,N,T,M <: Metric} <: NNTree{V,M}
-    data::Vector{V}
+    data::AbstractVector{V}
```

Comment:
This will be bad for performance because this is an abstract field. See (https://docs.julialang.org/en/stable/manual/performance-tips/#Avoid-containers-with-abstract-type-parameters-1),

Perhaps we need to just do like
 
```jl
BallTree{V <: AbstractVector, ...}
    data::V
    .
    .
end
```

Reply:
Indeed, this was a mistake on my part. I intended to leave `data::Vector{V}` there, like in `BruteTree`. In the latter it was necessary to do a `convert` in the `BruteTree` constructor, but not with `BallTree` I think.